src/FieldType/DBFocusPoint.php

```diff
@@ -0,0 +1,268 @@
+            // Focus point offset
+            $focusOffset = $this->focusCoordToOffset($cropAxis, $cropData[$cropAxis]['FocusPoint']);
+            // Length after scaling but before cropping
+            $scaledImageLength = floor($cropData[$cropAxis]['OriginalLength'] / $scaleRatio);
+            // Focus point position in pixels
+            $focusPos = floor($focusOffset * $scaledImageLength);
+            // Container center in pixels
+            $frameCenter = floor($cropData[$cropAxis]['TargetLength'] / 2);
+            // Difference beetween focus point and center
+            $focusShift = $focusPos - $frameCenter;
+            // Limit offset so image remains filled
+            $remainder = $scaledImageLength - $focusPos;
+            $croppedRemainder = $cropData[$cropAxis]['TargetLength'] - $frameCenter;
+            if ($remainder < $croppedRemainder) {
+                $focusShift -= $croppedRemainder - $remainder;
+            }
+            if ($focusShift < 0) {
+                $focusShift = 0;
+            }
+            // Set cropping start point
+            $cropData['CropOffset'] = $focusShift;
+            // Update Focus point location for cropped image
+            $newFocusOffset = ($focusPos - $focusShift) / $cropData[$cropAxis]['TargetLength'];
+            $cropData[$cropAxis]['FocusPoint'] = $this->focusOffsetToCoord($cropAxis, $newFocusOffset);
+        }
+
+        return $cropData;
+    }
+
+    /**
+     * Generate a cropped version of the given image
+     * @param int $width desired width
+     * @param int $height desired height
+     * @param Image $image the image to crop. If not set, the current record will be used
+     * @param bool $upscale whether or not upscaling is allowed
+     * @return AssetContainer|null
+     */
+    public function generateFocusFill($width, $height, Image $image, $upscale = true)
+    {
+        if (!$image && $this->record instanceof Image) {
+            $image = $this->record;
+        }
+        $width = intval($width);
+        $height = intval($height);
+        $imgW = $image->getWidth();
+        $imgH = $image->getHeight();
+
+        // Don't enlarge
+        if (!$upscale) {
+            $widthRatio = $imgW / $width;
+            $heightRatio = $imgH / $height;
+            if ($widthRatio < 1 && $widthRatio <= $heightRatio) {
+                $width = $imgW;
+                $height = intval(round($height * $widthRatio));
+            } elseif ($heightRatio < 1) {
+                $height = $imgH;
+                $width = intval(round($width * $heightRatio));
+            }
+        }
+
+        //Only resize if necessary
+        if ($image->isSize($width, $height) && !Config::inst()->get(DBFile::class, 'force_resample')) {
+            return $image;
+        } elseif ($cropData = $this->calculateCrop($width, $height, $imgW, $imgH)) {
+            $variant = $image->variantName(__FUNCTION__, $width, $height, $cropData['CropAxis'], $cropData['CropOffset']);
```

Comment:
This is a picky thing but now that this code is in a function called generateFocusFill, `__FUNCTION__` is returning that and so our resampled images are prefixed with 'generateFocusFill'. Can you replace `__FUNCTION__` here with 'FocusFill' please.

Reply:
Good point